### ISSUE TITLE
Rustdoc creates doctests with `extern crate mod;`

### ISSUE THREAD

**Issue by Swatinem** [rust-lang/rust#79771]
In #79770 I tried to create a new `rustdoc-ui` testcase that runs doctests. However having the `mod` substring in a doctest comment injected an invalid `extern crate mod;` into the code generated for the doctest.

Turns out this here is a bit too aggressive:

https://github.com/rust-lang/rust/blob/0f6f2d681b39c5f95459cd09cb936b6ceb27cd82/src/librustdoc/doctest.rs#L533-L534

It looks like compiletest actually gave the crate the name `mod`, perhaps because it was defined in a `issue-XXX/mod.rs` file, and having a doctest such as this triggers the this auto-injection, resulting in a failed doctest run:

````rust
/// ```
/// // ^ mod.rs line 10
/// assert_eq!(1, 1);
/// ```
````

**Comment by jyn514** [comment]
I think rustdoc needs to check if the crate name is a keyword and add `r#` if so. Actually it's probably enough to unconditionally add `r#` since it works for things that aren't keywords.

**Comment by Swatinem** [comment]
> Actually it's probably enough to unconditionally add `r#` since it works for things that aren't keywords.

there is actually two bugs here. `mod` not being properly escaped is one thing, the other is probably the way that rustdoc is invoked, or not quite sure. Basically it infers the crate name from the file name if not explicitly given, and then just assumes that a crate with that name exists, which it doesn’t.

**Comment by jyn514** [comment]
@Swatinem that's an issue with compiletest, not rustdoc - it needs to compile the crate itself before running rustdoc on it: https://github.com/rust-lang/rust/pull/79775#discussion_r537146641. I don't know how to fix that, I think compiletest needs to add support for this maybe?
